src/kv/test/kv_snapshot.cpp

```diff
@@ -0,0 +1,169 @@
+    {
+      REQUIRE_EQ(s->get_security_domain(), kv::SecurityDomain::PUBLIC);
+      REQUIRE_EQ(s->get_is_replicated(), true);
+
+      // Only string_map is committed at version 1
+      if (s->get_name() == "string_map")
+      {
+        REQUIRE_GT(s->get_serialized_size(), 0);
+      }
+      else
+      {
+        REQUIRE_EQ(s->get_name(), "num_map");
+        REQUIRE_EQ(s->get_serialized_size(), 0);
+      }
+    }
+  }
+
+  INFO("Apply snapshot at 1 to new store");
+  {
+    kv::Store new_store;
+    new_store.clone_schema(store);
+
+    new_store.deserialize(s_1);
+    REQUIRE_EQ(new_store.current_version(), 1);
+
+    auto new_string_map = new_store.get<MapTypes::StringString>("string_map");
+    auto new_num_map = new_store.get<MapTypes::NumNum>("num_map");
+
+    kv::Tx tx1;
+    auto view = tx1.get_view(*new_string_map);
+    auto v = view->get("foo");
+    REQUIRE(v.has_value());
+    REQUIRE_EQ(v.value(), "bar");
+
+    auto view_ = tx1.get_view(*new_num_map);
+    auto v_ = view_->get(42);
+    REQUIRE(!v_.has_value());
+
+    view = tx1.get_view(*new_string_map);
+    v = view->get("key");
+    REQUIRE(!v.has_value());
+  }
+
+  auto s_2 = store.snapshot(2);
+  INFO("Apply snapshot at 2 to new store");
+  {
+    kv::Store new_store;
+    new_store.clone_schema(store);
+    new_store.deserialize(s_2);
+    REQUIRE_EQ(new_store.current_version(), 2);
+
+    auto new_string_map = new_store.get<MapTypes::StringString>("string_map");
+    auto new_num_map = new_store.get<MapTypes::NumNum>("num_map");
+
+    kv::Tx tx1;
+    auto view = tx1.get_view(*new_string_map);
+
+    auto v = view->get("foo");
+    REQUIRE(v.has_value());
+    REQUIRE_EQ(v.value(), "bar");
+
+    auto view_ = tx1.get_view(*new_num_map);
+    auto v_ = view_->get(42);
+    REQUIRE(v_.has_value());
+    REQUIRE_EQ(v_.value(), 123);
+
+    view = tx1.get_view(*new_string_map);
+    v = view->get("key");
+    REQUIRE(!v.has_value());
+  }
+}
+
+TEST_CASE(
+  "Commit transaction while applying snapshot" *
+  doctest::test_suite("snapshot"))
+{
+  kv::Store store;
+  auto& string_map = store.create<MapTypes::StringString>(
+    "string_map", kv::SecurityDomain::PUBLIC);
+
+  INFO("Apply transactions to original store");
+  {
+    kv::Tx tx1, tx2;
+    auto view_1 = tx1.get_view(string_map);
+    auto view_2 = tx2.get_view(string_map);
+
+    view_1->put("foo", "foo");
+    view_2->put("bar", "bar");
+
+    REQUIRE(tx1.commit() == kv::CommitSuccess::OK); // Committed at 1
+    REQUIRE(tx2.commit() == kv::CommitSuccess::OK); // Committed at 2
+  }
+
+  auto s_1 = store.snapshot(2);
+
+  INFO("Apply snapshot while committing a transaction");
+  {
+    kv::Store new_store;
+    new_store.clone_schema(store);
+
+    auto new_string_map = new_store.get<MapTypes::StringString>("string_map");
+    kv::Tx tx1, tx2;
+    auto view = tx1.get_view(*new_string_map);
+    view->put("baz", "baz");
+    REQUIRE(tx1.commit() == kv::CommitSuccess::OK); // Committed at 1
```

Comment:
This seems like an extra complication we're not fully testing - we're deserialising a snapshot onto a kv which has a _different prefix_, it has had some totally different transactions executed on it before. It looks like this works (we drop the roll when applying a snapshot?), but we should test exactly this in isolation (ie - _this_ tx1 is totally lost after deserialising the snapshot)